modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryMemoryTracker.java

```diff
@@ -16,21 +16,33 @@
 
 package org.apache.ignite.internal.processors.query.h2;
 
-import org.apache.ignite.IgniteLogger;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
 import org.apache.ignite.internal.processors.cache.query.IgniteQueryErrorCode;
+import org.apache.ignite.internal.processors.query.GridQueryMemoryMetricProvider;
 import org.apache.ignite.internal.processors.query.IgniteSQLException;
+import org.apache.ignite.internal.util.tostring.GridToStringExclude;
 import org.apache.ignite.internal.util.typedef.internal.S;
 
 /**
  * Query memory tracker.
  *
  * Track query memory usage and throws an exception if query tries to allocate memory over limit.
  */
-public class QueryMemoryTracker implements H2MemoryTracker {
-    /** Logger. */
-    private final IgniteLogger log;
+public class QueryMemoryTracker implements H2MemoryTracker, GridQueryMemoryMetricProvider {
+    /** State updater. */
+    private static final AtomicIntegerFieldUpdater<QueryMemoryTracker> STATE_UPDATER
+        = AtomicIntegerFieldUpdater.newUpdater(QueryMemoryTracker.class, "state");
+
+    /** Tracker is not closed and not in the middle of the closing process. */
+    private static final int NORMALLY_OPERATING_STATE = 0;
+
+    /** Tracker is closed or in the middle of the closing process. */
+    private static final int CLOSING_OR_CLOSED_STATE = 1;
```

Comment:
What do you think about STATE_INIT, STATE_CLOSED ?

Reply:
ok